sct-commons/src/main/java/org/lfenergy/compas/sct/commons/ExtRefEditorService.java

```diff
@@ -126,83 +134,62 @@ private static Boolean doesExtRefMatchLDEPFChannel(TExtRef extRef, TChannel tCha
-                .map(compasICDHeader -> compasICDHeader.getIEDType().value().equals(channel.getIEDType())
+    private static boolean doesIcdHeaderMatchLDEPFChannel(TIED tied, TChannel channel) {
+        Optional<TCompasICDHeader> tCompasICDHeader = PrivateUtils.extractCompasPrivate(tied, TCompasICDHeader.class);
+        return tCompasICDHeader.map(compasICDHeader -> compasICDHeader.getIEDType().value().equals(channel.getIEDType())
                         && compasICDHeader.getIEDredundancy().value().equals(channel.getIEDRedundancy().value())
                         && compasICDHeader.getIEDSystemVersioninstance().toString().equals(channel.getIEDSystemVersionInstance()))
                 .orElse(false);
     }
 
-    /**
-     * Provides Active LDevice according to EPF channel's inst attribute
-     *
-     * @param iedAdapter IEDAdapter
-     * @param channel    TChannel
-     * @return LDeviceAdapter object that matches the EPF channel
-     */
-    private Optional<LDeviceAdapter> getActiveSourceLDeviceByLDEPFChannel(IEDAdapter iedAdapter, TChannel channel) {
-        return ldeviceService.findLdevice(iedAdapter.getCurrentElem(), channel.getLDInst())
-                .filter(tlDevice -> ldeviceService.getLdeviceStatus(tlDevice).map(ActiveStatus.ON::equals).orElse(false))
-                .map(tlDevice -> new LDeviceAdapter(iedAdapter, tlDevice));
-    }
-
     /**
      * Provides Active LN Object that satisfies the EPF channel attributes (lnClass, lnInst, prefix)
      *
-     * @param lDeviceAdapter LDeviceAdapter
+     * @param tlDevice TLDevice
      * @param channel        TChannel
      * @return AbstractLNAdapter object that matches the EPF channel
      */
-    private static Optional<AbstractLNAdapter<?>> getActiveLNSourceByLDEPFChannel(LDeviceAdapter lDeviceAdapter, TChannel channel) {
-        return lDeviceAdapter.getLNAdaptersIncludingLN0()
-                .stream()
-                .filter(lnAdapter -> lnAdapter.getLNClass().equals(channel.getLNClass())
-                        && lnAdapter.getLNInst().equals(channel.getLNInst())
-                        && trimToEmpty(channel.getLNPrefix()).equals(trimToEmpty(lnAdapter.getPrefix())))
-                .findFirst()
-                .filter(lnAdapter -> lnAdapter.getDaiModStValValue()
-                        .map(status -> status.equals(ActiveStatus.ON.getValue()))
-                        .orElse(true));
+    private Optional<TAnyLN> getActiveLNSourceByLDEPFChannel(TLDevice tlDevice, TChannel channel) {
+        return Stream.concat(Stream.of(tlDevice.getLN0()), tlDevice.getLN().stream())
+                .filter(tAnyLN -> {
+                    LnId lnId = LnId.from(tAnyLN);
+                    return lnId.lnClass().equals(channel.getLNClass())
+                            && lnId.lnInst().equals(channel.getLNInst())
+                            && trimToEmpty(channel.getLNPrefix()).equals(trimToEmpty(lnId.prefix()));
+                }).findFirst()
```

Comment:
You should put `LnService.getAnylns` and` LnService.matchesLn` in `LnEditor` interface and then you could write : 

```suggestion
        return lnEditor.getAnylns(tlDevice)
                .filter(tAnyLN -> lnEditor.matchesLn(tAnyLN, channel.getLNClass(), channel.getLNInst(), channel.getLNPrefix()))
                .findFirst()
```